### DISCUSSION TITLE
RenderErrorPage, 404 and onBeforeRender() in .page.js

### DISCUSSION THREAD

**Discussion by lutzissler** [vikejs/vike#718] [Discussion] [answered]
I have an app for which I provide all the necessary data in the first pageContext. After that, the app runs as an SPA and should therefore never call the server again. To avoid VPS doing those round trips to the server for new `pageContext`s, I have put `onBeforeRender()` in`.page.js` instead of `.page.server.js` [as the docs suggest](https://vite-plugin-ssr.com/onBeforeRender-isomorphic#recommendation).

However, now I’m struggling with issuing 404s for my dynamic routes. I did it as also [suggested by the docs](https://vite-plugin-ssr.com/route-function#redirection-404), i.e. by throwing a `RenderErrorPage()` in `onBeforeRender()` when I detect that the particular route doesn’t exist. However, that is only working as long as `onBeforeRender()` is in `.page.server.js`. As soon as I put it in `.page.js`, it won’t work because `RenderErrorPage()` is not available in the client and therefore `import { RenderErrorPage } from 'vite-plugin-ssr'` is failing on the client.

Am I trying the wrong thing? How else could I issue 404s in my server-side code while at the same time stopping VPS from calling the server for a `pageContext` on every navigation?

**Comment by brillout** [comment, marked as answer]
Check out https://github.com/brillout/vite-plugin-ssr/issues/717: you can now use `RenderErrorPage` from the client-side.

**Reply by lutzissler** [reply]
Just seen that! Awesome, thank you. Works like a charm.
